File: BinaryTreePath/BTP.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root: return []
        res = []
        self.subBTP(root,res,"")
        return res
        
    def subBTP(self, root: TreeNode, res, s):
        if not root.left and not root.right: res.append(s + str(root.val))
        if root.left:self.subBTP(root.left,res,s+str(root.val)+'->')
        if root.right:self.subBTP(root.right,res,s+str(root.val)+'->')

File: BinaryTreePath/BTP.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        ArrayList<String> res = new ArrayList<String>();
        if(root == null) return res ;
        subBTP(root,res,"");
        return res;
    }

    public void subBTP(TreeNode root, ArrayList<String> res, String s){
        if(root.left == null && root.right == null) res.add(s + root.val);
        if(root.left!= null) subBTP(root.left, res, s+root.val+"->");
        if(root.right!= null) subBTP(root.right, res, s+root.val+"->");
    }
}